### DISCUSSION TITLE
How to handle * route and which can return response or pass control to the next route

### DISCUSSION THREAD

**Discussion by Umenokin** [tokio-rs/axum#2276] [Q&A] [answered]
I am trying to provide CMS-like routing like:

1. `/* - user pages`
2. `/admin/* - admin dashboard`

## My current solution

While ideally, I would prefer to check user pages first I managed to make a working example based on Leptos's examples where:
1. Check admin (Leptos) routes
2. If not Leptos routes, got to fallback
3. Return admin file asset if one was found
4. Check the user page (`runtime_handler` is the one serving those)
5. Return error

```
// Routes
    let app = Router::new()
        .leptos_routes(&leptos_options, routes, App)
        .fallback(runtime_and_other_handlers)
        .with_state(leptos_options);
        
// Fallback
pub async fn runtime_and_other_handlers(
    uri: Uri,
    State(leptos_options): State<LeptosOptions>,
    req: Request<Body>,
) -> AxumResponse {
    let root = leptos_options.site_root.clone();
    let res = get_static_file(uri.clone(), &root).await.unwrap();

    if res.status() == StatusCode::OK {
        res.into_response()
    } else {
        let res = runtime_handler(&req).await;
        if res.status() == StatusCode::OK {
            return res;
        }

        let mut errors = Errors::default();
        errors.insert_with_default_key(AppError::NotFound);
        let handler = leptos_axum::render_app_to_stream(
            leptos_options.to_owned(),
            move || view! { <ErrorTemplate outside_errors=errors.clone()/> },
        );
        handler(req).await.into_response()
    }
}
```
The problem with this approach is that I spend time checking first admin and then admin assets before doing actual work while adding is something that should be rarely used:

It would be great if I could do it in the following order:
1. Check the user page (`runtime_handler` is the one serving those)
2. If any of `runtime_handler` routes matched the return response
2. If not, check the rest of the routes
3. Go to fallback

In Axum terms, it would look something like this:
```
app
  .route("*", any(
    const response =  check_crms_routes()
    if response {
      return response
    } else {
      // check rest of the axum routes
    }
  ))
  .leptos_routes(&leptos_options, routes, App)
  .fallback(files_and_error_handlers)
```

**Comment by davidpdrsn** [comment]
Doesnt

```
.route("/*path", ...).route("/admin/*path", ...)
```

work?

**Reply by Umenokin** [reply]
Thank you @davidpdrsn. You as always very quick and responsive. Wonder, why you still did not switch the "sponsoring" feature on 😅

Unfortunately, it works only partially `/*path` will intercept all paths, including static assets. But I think I know what I should do which is making a bit smarter leptos routing something like below, correct?
```
.route("/public/*path", get_static_files)
.route("/*path", get_static_files)
.route("/admin)
```

Or to make sure that all Leptos assets are also located  within `admin/*` space

Though point me to the doc explaining how routes are evaluated. I think there was one which it seems NOT this one [routing](https://docs.rs/axum/latest/axum/#routing). I remember something like "routes are evaluated bottom-top" otherwise I will be confused how "/admin/*" is detected while route `/*path` will match everything.

**Reply by jplatte** [reply]
I'm not sure where this is documented, but *more specific* paths take precedence over less specific ones. That is you can have both `/:var` and `/static_segment`, and the latter will take precedence, and you can have `/*path` and `/foo/*path` and `/bar/*path` and the first only matches if the other two don't match. The order in which the routes are added doesn't matter. If the routes are truly ambiguous (none is more specific than the other and they overlap), building the `Router` will panic.

**Reply by davidpdrsn** [reply]
Most of the routing options are here https://docs.rs/axum/latest/axum/struct.Router.html#method.route

> I remember something like "routes are evaluated bottom-top"

This was the case once but hasn’t been true for a long time. These days the ordering doesn’t matter.

> But I think I know what I should do which is making a bit smarter leptos routing something like below, correct?

Yes I’d recommend a setup like that.

**Reply by Umenokin** [reply, marked as answer]
Thank you very much @davidpdrsn, @jplatte.
It seems I managed to make a solution that works well for now:

```
let app = Router::new()
        .route("/*path", any(|| async { Html("Hello, World!") }))
        .route("/pkg/*path", get(handle_static_files))
        .route("/admin/assets/*path", get(handle_static_files))
        .leptos_routes(&leptos_options, routes, App)
        .with_state(leptos_options);
```

I would've wanted to move pkg under `admin` for consistency but this is purely a `leptos` configuration.

Thanks a lot!
